File: backend/app/routers/room_types.py
from datetime import date, timedelta

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db import get_session
from app.models import Availability, RatePlan, RoomType
from app.schemas.public import AvailabilityDayOut, RatePlanOut, RoomAvailabilityOut

router = APIRouter(prefix="/api/room-types", tags=["room-types"])

MAX_AVAILABILITY_RANGE_DAYS = 60


@router.get("/{room_type_id}", response_model=RoomType)
async def get_room_type(room_type_id: int, session: AsyncSession = Depends(get_session)):
    room_type = await session.get(RoomType, room_type_id)
    if not room_type:
        raise HTTPException(404, "Room type not found")
    return room_type


@router.get("/{room_type_id}/availability", response_model=RoomAvailabilityOut)
async def get_availability(
    room_type_id: int,
    checkin: date,
    checkout: date,
    session: AsyncSession = Depends(get_session),
):
    if checkin >= checkout:
        raise HTTPException(400, "checkout must be after checkin")
    if (checkout - checkin).days > MAX_AVAILABILITY_RANGE_DAYS:
        raise HTTPException(400, f"Date range cannot exceed {MAX_AVAILABILITY_RANGE_DAYS} days")

    days = [checkin + timedelta(days=i) for i in range((checkout - checkin).days)]
    avail_result = await session.execute(
        select(Availability).where(
            Availability.room_type_id == room_type_id, Availability.date.in_(days)
        )
    )
    by_date = {row.date: row.rooms_available for row in avail_result.scalars().all()}

    rates_result = await session.execute(
        select(RatePlan).where(
            RatePlan.room_type_id == room_type_id,
            RatePlan.valid_from <= checkin,
            RatePlan.valid_to >= checkout,
        )
    )
    rate_plans = rates_result.scalars().all()

    return RoomAvailabilityOut(
        room_type_id=room_type_id,
        days=[AvailabilityDayOut(date=d, rooms_available=by_date.get(d, 0)) for d in days],
        rate_plans=[
            RatePlanOut(
                id=r.id,
                name=r.name,
                fixed_price=float(r.fixed_price),
                refundable=r.refundable,
                includes_breakfast=r.includes_breakfast,
            )
            for r in rate_plans
        ],
    )

File: backend/src/routes/roomTypes.ts
import { Router } from "express";
import { query } from "../db";
import { asyncHandler, HttpError } from "../middleware/errors";
import { Availability, RatePlan, RoomType } from "../types/models";

export const router = Router();

const MAX_AVAILABILITY_RANGE_DAYS = 60;
const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

function addDays(dateStr: string, days: number): string {
  const d = new Date(dateStr + "T00:00:00Z");
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

function daysBetween(a: string, b: string): number {
  return Math.round((Date.parse(b + "T00:00:00Z") - Date.parse(a + "T00:00:00Z")) / 86_400_000);
}

router.get(
  "/:roomTypeId",
  asyncHandler(async (req, res) => {
    const roomTypeId = parseInt(req.params.roomTypeId, 10);
    const rows = await query<RoomType>("SELECT * FROM room_types WHERE id = $1", [roomTypeId]);
    const roomType = rows[0];
    if (!roomType) throw new HttpError(404, "Room type not found");
    res.json(roomType);
  })
);

router.get(
  "/:roomTypeId/availability",
  asyncHandler(async (req, res) => {
    const roomTypeId = parseInt(req.params.roomTypeId, 10);
    const checkin = String(req.query.checkin ?? "");
    const checkout = String(req.query.checkout ?? "");
    if (!DATE_RE.test(checkin) || !DATE_RE.test(checkout)) throw new HttpError(422, "Invalid date format, expected YYYY-MM-DD");
    if (checkin >= checkout) throw new HttpError(400, "checkout must be after checkin");
    if (daysBetween(checkin, checkout) > MAX_AVAILABILITY_RANGE_DAYS) {
      throw new HttpError(400, `Date range cannot exceed ${MAX_AVAILABILITY_RANGE_DAYS} days`);
    }

    const days: string[] = [];
    for (let i = 0; i < daysBetween(checkin, checkout); i++) days.push(addDays(checkin, i));

    const availRows = await query<Availability>(
      "SELECT * FROM availability WHERE room_type_id = $1 AND date = ANY($2::date[])",
      [roomTypeId, days]
    );
    const byDate = new Map(availRows.map((r) => [r.date, r.rooms_available]));

    const ratePlans = await query<RatePlan>(
      "SELECT * FROM rate_plans WHERE room_type_id = $1 AND valid_from <= $2 AND valid_to >= $3",
      [roomTypeId, checkin, checkout]
    );

    res.json({
      room_type_id: roomTypeId,
      days: days.map((d) => ({ date: d, rooms_available: byDate.get(d) ?? 0 })),
      rate_plans: ratePlans.map((r) => ({
        id: r.id,
        name: r.name,
        fixed_price: r.fixed_price,
        refundable: r.refundable,
        includes_breakfast: r.includes_breakfast,
      })),
    });
  })
);
